api/auth/gcp/gcp.go

```diff
@@ -0,0 +1,184 @@
+package gcp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"cloud.google.com/go/compute/metadata"
+	credentials "cloud.google.com/go/iam/credentials/apiv1"
+	"github.com/hashicorp/vault/api"
+	credentialspb "google.golang.org/genproto/googleapis/iam/credentials/v1"
+)
+
+type GCPAuth struct {
+	roleName            string
+	mountPath           string
+	authType            string
+	serviceAccountEmail string
+}
+
+var _ api.AuthMethod = (*GCPAuth)(nil)
+
+type LoginOption func(a *GCPAuth) error
+
+const (
+	iamType             = "iam"
+	gceType             = "gce"
+	defaultMountPath    = "gcp"
+	defaultAuthType     = gceType
+	identityMetadataURL = "http://metadata/computeMetadata/v1/instance/service-accounts/default/identity"
+)
+
+// NewGCPAuth initializes a new GCP auth method interface to be
+// passed as a parameter to the client.Auth().Login method.
+//
+// Supported options: WithMountPath, WithIAMAuth, WithGCEAuth
+func NewGCPAuth(roleName string, opts ...LoginOption) (*GCPAuth, error) {
+	if roleName == "" {
+		return nil, fmt.Errorf("no role name provided for login")
+	}
+
+	a := &GCPAuth{
+		mountPath: defaultMountPath,
+		authType:  defaultAuthType,
+		roleName:  roleName,
+	}
+
+	// Loop through each option
+	for _, opt := range opts {
+		// Call the option giving the instantiated
+		// *GCPAuth as the argument
+		err := opt(a)
+		if err != nil {
+			return nil, fmt.Errorf("error with login option: %w", err)
+		}
+	}
+
+	// return the modified auth struct instance
+	return a, nil
+}
+
+// Login sets up the required request body for the GCP auth method's /login
+// endpoint, and performs a write to it. This method defaults to the "gce"
+// auth type unless NewGCPAuth is called with WithIAMAuth().
+func (a *GCPAuth) Login(ctx context.Context, client *api.Client) (*api.Secret, error) {
+	loginData := map[string]interface{}{
+		"role": a.roleName,
+	}
+	switch a.authType {
+	case gceType:
+		jwt, err := a.getJWTFromMetadataService(client.Address())
+		if err != nil {
+			return nil, fmt.Errorf("unable to retrieve JWT from GCE metadata service: %w", err)
+		}
+		loginData["jwt"] = jwt
+	case iamType:
+		jwtResp, err := a.signJWT()
+		if err != nil {
+			return nil, fmt.Errorf("unable to sign JWT for authenticating to GCP: %w", err)
+		}
+		loginData["jwt"] = jwtResp.SignedJwt
+	}
+
+	path := fmt.Sprintf("auth/%s/login", a.mountPath)
+	resp, err := client.Logical().Write(path, loginData)
+	if err != nil {
+		return nil, fmt.Errorf("unable to log in with GCP auth: %w", err)
+	}
+
+	return resp, nil
+}
+
+func WithMountPath(mountPath string) LoginOption {
+	return func(a *GCPAuth) error {
+		a.mountPath = mountPath
+		return nil
+	}
+}
+
+func WithIAMAuth(serviceAccountEmail string) LoginOption {
```

Comment:
It'd be cool to also have an option to just supply a JWT. For example, you can obtain a signed JWT using gcloud or curl ([examples](https://www.vaultproject.io/docs/auth/gcp#curl-example)). Not a blocker though, can always add this later.